### PR TITLE
Development monitoring

### PR DESCRIPTION
**Pull request by 0oM4R**
### Description

This package provides service monitoring; currently, check if the service is alive 

#### `ServiceMonitor` class
  - `checkServiceAliveness` : accepts array of `IServiceAliveness` and 
   Creates an instance of ServiceMonitor.
     - services - An array of services to monitor.
     - interval - The interval, in minutes, between monitoring checks (default is 2 minutes)
     -  retries - The number of retries in case a service is determined to be down (default is 2 retries).
     - retryInterval - The interval, in seconds, between retries (default is 2 seconds).
   This class have three public functions 

    1. `monitorService` : Monitors the services at a regular interval and returns a function to exit and disconnect the monitoring.
    2. `pingService`: check the liveness of the services and disconnect 
    3. `disconnect`: disconnect the services that have disconnect handler

  

#### `ILivenessChecker` Interface
each instance of the classes that implements `ILivenessChecker` interface have its own logic to create the connection and 
check the service's aliveness 

#### Evenets 
we expose an instance of `MonitorEventEmitter` that make emitting events more easier 

Screenshot from 2024-01-22 16-39-21

### Changes 

added new package `monitoring`

### Related Issues

- #1469

### Checklist

- [ ] Tests included
- [x] Build pass
- [x] Documentation
- [x] Code format and docstrings
- [x] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

**Comment by MohamedElmdary** [comment]
```ts
export abstract class MonitorBase {
  constructor(protected readonly serviceName: string) {}

  public abstract isAlive(): Promise<ServiceStatus>;
}
```

Why don't we use `MonitorBase` abstract class and use it everyWhere instead of using `serviceName` in interface?

File: packages/monitoring/package.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "monitoring",
+  "version": "1.0.0",
+  "description": "Threefold monitoring package",
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "example": "yarn ts-node ./example/index.js",
```
Comment (MohamedElmdary) [inline packages/monitoring/package.json:8]:
```suggestion
    "example": "yarn ts-node ./example/index.ts",
```

Example should be executed directly from typescript code

File: packages/monitoring/src/types/index.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  Black = "\x1b[30m",
+  Red = "\x1b[31m",
+  Green = "\x1b[32m",
+  Yellow = "\x1b[33m",
+  Blue = "\x1b[34m",
+  Magenta = "\x1b[35m",
+  Cyan = "\x1b[36m",
+  White = "\x1b[37m",
+  Gray = "\x1b[90m",
+}
```
Comment (MohamedElmdary) [inline packages/monitoring/src/types/index.ts:28]: Checkout https://www.npmjs.com/package/chalk
Chalk is doing the same functionality without any configuration

File: packages/monitoring/src/serviceMonitor/graphql.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      return {
+        alive: true,
+      };
+    } catch (error) {
+      return {
+        alive: false,
+        error,
+      };
+    }
+  }
```
Comment (MohamedElmdary) [inline packages/monitoring/src/serviceMonitor/graphql.ts:22]:
```ts
// utils.ts
// this is a util method which handle returned status
// so we can update how return status look from a single method
export async function resolveServiceStatus(promise: Promise<any>) {
  try {
    await promise;
    return { alive: true };
  } catch (error) {
    return { alive: false, error };
  }
}
```

```suggestion
  isAlive(): Promise<ServiceStatus> {
     return resolveServiceStatus(sendGetRequest(this.ServiceURL, "", {}))
   }
```

File: packages/monitoring/src/helpers/utils.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { RequestError } from "@threefold/types";
+import axios, { AxiosError } from "axios";
+
+import { TextColor } from "../types";
+
+export async function sendGetRequest(url: string, body: string, headers: Record<string, string>) {
```
Comment (MohamedElmdary) [inline packages/monitoring/src/helpers/utils.ts:6]: Get request doesn't have body ... also we don't need any headers here for now so we can mark headers as optional

File: packages/monitoring/src/types/index.ts (hunk 2 above)
Comment (0oM4R) [reply packages/monitoring/src/types/index.ts]: awesome , thanks

File: packages/monitoring/src/helpers/utils.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    const { response } = e as AxiosError;
+    let errorMessage = (response?.data as { error: string })?.error;
+    if (!errorMessage) {
+      errorMessage = (e as AxiosError).message;
+    }
+    throw new RequestError(`HTTP request failed ${errorMessage ? "due to " + errorMessage : ""}.`);
+  }
+}
+
+export function generateString(length: number): string {
```
Comment (xmonader) [inline packages/monitoring/src/helpers/utils.ts:23]: Why didn't you include tests for this?

File: packages/monitoring/src/helpers/utils.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import { ServiceStatus } from "src/types";
+
+export async function sendGetRequest(url: string, headers?: Record<string, string>) {
+  const options = {
+    method: "get",
+    url: url,
+    headers: headers || {},
+  };
+  try {
+    return await axios(options);
```
Comment (xmonader) [inline packages/monitoring/src/helpers/utils.ts:12]: why not axios.get directly?

File: packages/monitoring/src/helpers/utils.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import { RequestError } from "@threefold/types";
+import axios, { AxiosError } from "axios";
+import { ServiceStatus } from "src/types";
+
+export async function sendGetRequest(url: string, headers?: Record<string, string>) {
```
Comment (xmonader) [inline packages/monitoring/src/helpers/utils.ts:5]: why didn't you default the headers to {} ?

File: packages/monitoring/src/helpers/utils.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  const options = {
+    method: "get",
+    url: url,
+    headers: headers || {},
+  };
+  try {
+    return await axios(options);
+  } catch (e) {
+    const { response } = e as AxiosError;
+    let errorMessage = (response?.data as { error: string })?.error;
```
Comment (xmonader) [inline packages/monitoring/src/helpers/utils.ts:15]: this should have been more concise e.response?.data.error || e.message I believe

File: packages/monitoring/src/helpers/events.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import chalk from "chalk";
+import { EventEmitter } from "events";
+
+const events = new EventEmitter();
+type ServiceStatus = { [key: string]: boolean };
+
+const serviceStatusSummary: ServiceStatus = {};
+const ALIVE = chalk.green.bold("Alive");
+const DOWN = chalk.red.bold("Down");
+function logsHandler(msg) {
```
Comment (xmonader) [inline packages/monitoring/src/helpers/events.ts:10]: won't this duplicate the logs coming from the grid client? I thought we were going to create another sink

File: packages/monitoring/src/helpers/events.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+function printStatusSummary() {
+  const serviceNames = Object.keys(serviceStatusSummary);
+  const maxServiceNameLength = Math.max(...serviceNames.map(entry => entry.length));
+  console.log(chalk.blue.bold("Aliveness check summary:"));
+  for (const service in serviceStatusSummary) {
+    const padding = " ".repeat(maxServiceNameLength - service.length);
+    console.log(`\t${service}${padding}: ${serviceStatusSummary[service] ? ALIVE : DOWN}`);
+  }
+}
+events.addListener("logs", logsHandler);
```
Comment (xmonader) [inline packages/monitoring/src/helpers/events.ts:33]: and in the dashboard package ppl should listen to these events to provide custom callbacks in the UI?

File: packages/monitoring/src/serviceMonitor/disconnectHandler.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { events } from "../helpers/events";
+import { RMBMonitor, TFChainMonitor } from "../index";
+import { IServiceAliveness } from "../types";
+export async function disconnectServices(services: IServiceAliveness[]): Promise<void> {
+  for (const service of services) {
+    if (service instanceof RMBMonitor || service instanceof TFChainMonitor) {
```
Comment (xmonader) [inline packages/monitoring/src/serviceMonitor/disconnectHandler.ts:6]: The check should be if the service implements the IDisconnector

File: packages/monitoring/src/serviceMonitor/alivenessChecker.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+ * @throws Will throw an error if the provided services array is empty.
+ *
+ * @returns An object with a method `exitAndDisconnect` that can be used to stop the monitoring and disconnect services.
+ *
+ */
+export function monitorServiceAliveness(
+  services: IServiceAliveness[],
+  interval = 2,
+): { exitAndDisconnect: () => Promise<void> } {
+  if (!services.length) throw new Error("No services to monitor");
```
Comment (xmonader) [inline packages/monitoring/src/serviceMonitor/alivenessChecker.ts:20]: be careful with your checks, do services.length === 0, while your code works, still better to be as explicit as possible

File: packages/monitoring/src/serviceMonitor/alivenessChecker.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
+import { events } from "../helpers/events";
```
Comment (xmonader) [inline packages/monitoring/src/serviceMonitor/alivenessChecker.ts:1]: I don't uderstand why we didn't do it like this


class ServiceMonitor {

    constructor(services: Array<IService>, interval) {
        ...
    }
    private checkLivenessOnce {

    }
    private checkHealthOnce {

    }
    monitor() {
        withInterval do your checkLivenessOnce and checkHealthOnce 
    }

}

File: packages/monitoring/src/helpers/events.ts (hunk 9 above)
Comment (0oM4R) [reply packages/monitoring/src/helpers/events.ts:10]: will add prefix to the event names

File: packages/monitoring/example/index.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import { GridProxyMonitor, RMBMonitor, ServiceMonitor, TFChainMonitor } from "../src/";
+async function HealthCheck() {
+  try {
+    const services = [
+      new GridProxyMonitor("<FakeURL>"),
+      new TFChainMonitor("wss://tfchain.dev.grid.tf/ws", "mnemonic", "sr25519"),
+      new RMBMonitor("wss://relay.dev.grid.tf", "wss://tfchain.dev.grid.tf/ws", "mnemonic", "sr25519"),
+    ];
```
Comment (mohamedamer453) [inline packages/monitoring/example/index.ts:8]: Would be great if a check for graphql was added to the example.

**Review by mohamedamer453 — APPROVED** [review]

File: packages/monitoring/src/serviceMonitor/tfChain.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+import { IDisconnectHandler, ILivenessChecker, ServiceStatus } from "../types";
+
+export class TFChainMonitor implements ILivenessChecker, IDisconnectHandler {
+  private name = "TFChain";
+  private url: string;
+  private tfClient: TFClient;
+  constructor(tfChainUrl: string, mnemonic: string, keypairType: KeypairType) {
+    this.url = tfChainUrl;
+    this.tfClient = new TFClient({ url: this.url, mnemonicOrSecret: mnemonic, keypairType: keypairType });
```
Comment (AhmedHanafy725) [inline packages/monitoring/src/serviceMonitor/tfChain.ts:12]: why are we using TFClient here? it requires a mnemonic and a keypair type. Instead, we can use the queryClient which requires only the url

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]
I am afraid about disconnecting the clients cause it's only one static instance per url on the client, so if it will be used in the dashboard it might disconnect the client while deploying something.

File: packages/monitoring/package.json [hunk 16]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "monitoring",
```
[the remark this answers is not in the document]
Comment (0oM4R) [reply packages/monitoring/package.json]: i notice the license also is missing, which license should we use ?

**Pull request merged**